src/python/pants/engine/subsystem/native.py

```diff
@@ -299,20 +300,49 @@ class RunnableComplete(datatype('RunnableComplete', ['value', 'is_throw'])):
   """Corresponds to the native object of the same name."""
 
 
-class ExternContext(object):
-  """A wrapper around python objects used in static extern functions in this module.
+class ObjectIdMap(object):
+  """In the native context, assign and memoize python objects an unique unsigned-integer Id.
+
+  Underlying, the id is uniquely derived from object's digest instead of using its hash code
+  because the content may change and object's hash function could be overridden. In order not
+  to return the stale object we trade performance for correctness.
 
-  In the native context, python objects are identified by an unsigned-integer Id which is
-  assigned and memoized here. Note that this is independent-from and much-lighter-than
-  the Digest computed when an object is stored via storage.py (which is generally only necessary
-  for multi-processing or cache lookups).
+  In the future, we could improve performance by only computing digests for mutable objects.
+  For this reason referring an implementation-independent id instead of the digest in the native
+  context is more flexible.
   """
 
   def __init__(self):
+    # objects indexed by their keys, i.e, content digests
+    self._objects = Storage.create(in_memory=True)
+    # Memoized object Ids.
+    self._id_to_key = dict()
+    self._key_to_id = dict()
+    self._next_id = 0
```

Comment:
The difference is `_next_id` allows itself be only conditionally incremented for *new* objects. Using `_next_id` is actually 3 steps: 
- get next_id, 
- check if already used, 
- may or may not increment.